### PR TITLE
make exp nothrow

### PR DESCRIPTION
**Pull request by oscardssmith**
before it wasn't because the compiler can't prove that the getindex into the table is inbounds.

### CODE REVIEW

**Review by aviatesk — APPROVED** [review]
Add a test case that shows the improvement.

File: test/math.jl [hunk 1]
```diff
@@ -1461,7 +1461,11 @@ for fn in (:sin, :cos, :tan, :log, :log2, :log10, :log1p, :exponent, :sqrt, :cbr
         @test Core.Compiler.is_foldable(eff)
     end
 end
-for T in (Float32, Float64)
+for T in (Float16, Float32, Float64)
+    for f in (exp, exp2, exp10)
+        @test Base.infer_effects(f, T).nothrow
+    end
+    @test Core.Compiler.is_foldable(Base.infer_effects(, (T,T)))
```
Comment (simeonschaub) [inline test/math.jl:1468]:
```suggestion
    @test Core.Compiler.is_foldable(Base.infer_effects(^, (T,T)))
```

File: test/math.jl [hunk 2]
```diff
@@ -1461,7 +1461,10 @@ for fn in (:sin, :cos, :tan, :log, :log2, :log10, :log1p, :exponent, :sqrt, :cbr
         @test Core.Compiler.is_foldable(eff)
     end
 end
-for T in (Float32, Float64)
+for T in (Float16, Float32, Float64)
+    for f in (exp, exp2, exp10)
+        @test Base.infer_effects(f, (T,)).nothrow
```
Comment (aviatesk) [inline test/math.jl:1466]:
```suggestion
        @test Core.Compiler.is_removable_if_unused(Base.infer_effects(f, (T,)))
```

**Review by aviatesk — COMMENTED** [review]
We generall want to avoid accessing internals of `Effects` but use high-level queries instead, as the data structure isn't very stable yet.

**Comment by ianatol** [comment]
Related: https://github.com/JuliaLang/julia/pull/46199

**Comment by oscardssmith** [comment]
That's not related. That's about getfield analysis of mutable objects.

**Comment by ianatol** [comment]
Oops, my mistake. I had thought the improved `:consistency` analysis of `getfield` was what allowed us to make this change --- didn't realize it wasn't applicable here

**Pull request merged**